### PR TITLE
New in town challenges

### PR DESCRIPTION
**Pull request by Catalin-Gruia**
Pull request for "New in town" challenges.

### CODE REVIEW

File: src/test/kotlin/com/igorwojda/integer/addupto/challenge.kt [hunk 1]
```diff
@@ -4,7 +4,8 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun addUpTo(n: Int): Int {
-    TODO("not implemented")
+    // Gauss formula for consecutive numbers 1 + 2 + 3 + 4 + … + n == n x ( n + 1 ) : 2
+    return n*(n+1)/2
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/integer/addupto/challenge.kt]: nitpick: this is indeed the most optimal solution. I'd also suggest looking into aggregate operations: https://kotlinlang.org/docs/collection-aggregate.html

File: src/test/kotlin/com/igorwojda/integer/factorial/challenge.kt [hunk 2]
```diff
@@ -4,7 +4,10 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun factorial(n: Int): Int {
-    TODO("not implemented")
+    var total = 1
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/integer/factorial/challenge.kt:8]: Re-write this:
1. Using a folding operation
2. Using a recursive algorithm
3. If you're curious, look at tail recursive functions (https://kotlinlang.org/docs/reference/functions.html#tail-recursive-functions). See if you can write factorial as a tailrec function.
Please compare the running time of the solutions.

File: src/test/kotlin/com/igorwojda/integer/power/challenge.kt [hunk 3]
```diff
@@ -2,9 +2,10 @@ package com.igorwojda.integer.power
 
 import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
+import kotlin.math.pow
 
 private fun power(base: Int, exponent: Int): Int {
-    TODO("not implemented")
+    return base.toDouble().pow(exponent.toDouble()).toInt()
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/integer/power/challenge.kt]: Challenge: reimplement this by using the following concepts:
1. List initializer: https://kotlinlang.org/docs/reference/constructing-collections.html#initializer-functions-for-lists
2. Reduce operation: https://kotlinlang.org/docs/reference/collection-aggregate.html#fold-and-reduce
3. An operator: https://kotlinlang.org/docs/reference/operator-overloading.html
4. A method reference (to the above operator): https://khan.github.io/kotlin-for-python-developers/#function-references

It should be a one-liner ;)

File: src/test/kotlin/com/igorwojda/list/capitalizeFirst/challenge.kt [hunk 4]
```diff
@@ -4,7 +4,9 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun capitalizeFirst(list: List<String>): List<String> {
-    TODO("not implemented")
+    return if (list.isEmpty())
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/capitalizeFirst/challenge.kt]: Please look at collection transformation operations: https://kotlinlang.org/docs/collection-transformations.html

File: src/test/kotlin/com/igorwojda/list/countuniquevalues/challenge.kt [hunk 5]
```diff
@@ -4,7 +4,7 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun countUniqueValues(list: List<Int>): Int {
-    TODO("not implemented")
+    return list.distinct().size
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/countuniquevalues/challenge.kt]: challenge: convert the List to a different type of collection that yields the unique items in the list. Can you think which one it is?

File: src/test/kotlin/com/igorwojda/list/formattrainroute/challenge.kt [hunk 6]
```diff
@@ -4,7 +4,22 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun formatTrainRoute(stations: List<String>): String {
-    TODO("not implemented")
+    if (stations.size > 1) {
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/formattrainroute/challenge.kt:7]: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html is a very powerful tool

File: src/test/kotlin/com/igorwojda/list/product/challenge.kt [hunk 7]
```diff
@@ -4,7 +4,11 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun product(list: List<Int>): Int {
-    TODO("not implemented")
+    var product = 1
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/product/challenge.kt]: Challenge: reimplement this by using the following concepts:
1. Reduce operation: https://kotlinlang.org/docs/reference/collection-aggregate.html#fold-and-reduce
2. An operator: https://kotlinlang.org/docs/reference/operator-overloading.html
3. A method reference (to the above operator): https://khan.github.io/kotlin-for-python-developers/#function-references
It should be a single chain of functional calls.

File: src/test/kotlin/com/igorwojda/list/search/linearsearch/challenge.kt [hunk 8]
```diff
@@ -4,7 +4,10 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun getIndex(list: List<String>, str: String): Int {
-    TODO("not implemented")
+    for (item in list.indices)
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/search/linearsearch/challenge.kt]: nitpick: there's also a forEachIndexed on collections

**Review by radu-marin-sv — CHANGES_REQUESTED** [review]
Before addressing the comments, please check out https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods, especially for formatting and indentation

File: src/test/kotlin/com/igorwojda/integer/power/challenge.kt [hunk 9]
```diff
@@ -4,7 +4,12 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun power(base: Int, exponent: Int): Int {
-    TODO("not implemented")
+    // Implementation below does the same thing, but with less code
+    //    return List(exponent) { base }.reduce { product, element -> product * element }
+
+    // But since you asked to have all the required concepts, please check the implementation below :)
+    operator fun List<Int>.invoke() = reduce { it, product -> it * product }
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/integer/power/challenge.kt:11]: Why is this line required?

File: src/test/kotlin/com/igorwojda/list/product/challenge.kt [hunk 10]
```diff
@@ -4,7 +4,12 @@ import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 
 private fun product(list: List<Int>): Int {
-    TODO("not implemented")
+    // Implementation below does the same thing, but with less code
+    //    return list.reduce { it, product -> product * it }
+
+    // But since you asked to have all the required concepts, please check the implementation below :)
+    operator fun List<Int>.invoke() = reduce { it, product -> it * product }
```
Comment (radu-marin-sv) [inline src/test/kotlin/com/igorwojda/list/product/challenge.kt:11]: Why is this line required?

**Review by radu-marin-sv — APPROVED** [review]

File: src/test/kotlin/com/igorwojda/integer/power/challenge.kt (hunk 9 above)
Comment (Catalin-Gruia) [reply src/test/kotlin/com/igorwojda/integer/power/challenge.kt:11]: Yes, you are right. It is not required, I can implement this challenge with only the following line:

` return List(exponent) { base }.reduce(Int::times)`

File: src/test/kotlin/com/igorwojda/list/product/challenge.kt (hunk 10 above)
Comment (Catalin-Gruia) [reply src/test/kotlin/com/igorwojda/list/product/challenge.kt:11]: Yes, you are right. It is not required, I can implement this challenge with only the following line:

`return list.reduce(Int::times)`

**Review by radu-marin-sv — APPROVED** [review]

**Comment by radu-marin-sv** [comment]
Please use "Squash and merge"
